File: src/main/java/xyz/oribuin/skyblock/command/impl/TeleportCommand.java
package xyz.oribuin.skyblock.command.impl

import dev.rosewood.rosegarden.RosePlugin
import dev.rosewood.rosegarden.command.framework.CommandContext
import dev.rosewood.rosegarden.command.framework.RoseCommand
import dev.rosewood.rosegarden.command.framework.RoseCommandWrapper
import dev.rosewood.rosegarden.command.framework.annotation.Optional
import dev.rosewood.rosegarden.command.framework.annotation.RoseExecutable
import org.bukkit.entity.Player
import xyz.oribuin.skyblock.manager.IslandManager
import xyz.oribuin.skyblock.util.asMember
import xyz.oribuin.skyblock.util.asPlayer
import xyz.oribuin.skyblock.util.getManager

class TeleportCommand(rosePlugin: RosePlugin, parent: RoseCommandWrapper) : RoseCommand(rosePlugin, parent) {

    @RoseExecutable
    fun execute(context: CommandContext, @Optional player: Player?) {

        val manager = this.rosePlugin.getManager<skyblock.manager.IslandManager>()
        val target = (player ?: context.asPlayer()).asMember(this.rosePlugin)
        val island = manager.getIsland(target)

        if (island == null) {
            this.rosePlugin.getManager<skyblock.manager.MenuManager>()[xyz.oribuin.skyblock.gui.CreateGUI::class].openMenu(target)
            return
        }

        manager.teleport(target, island.home)

    }

    override fun getDefaultName(): String = "teleport"

    override fun getDefaultAliases(): List<String> = listOf("tp", "home", "go", "spawn")

    override fun getDescriptionKey(): String = "command-teleport-description"

    override fun getRequiredPermission() = "skyblock..teleport"

    override fun isPlayerOnly(): Boolean = true

}

File: Plugin/src/main/kotlin/xyz/golimc/skyblock/command/TeleportCommand.kt
package xyz.golimc.skyblock.command

import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import xyz.golimc.skyblock.SkyblockPlugin
import xyz.golimc.skyblock.manager.DataManager
import xyz.golimc.skyblock.manager.IslandManager
import xyz.golimc.skyblock.util.getManager
import xyz.golimc.skyblock.util.send
import xyz.oribuin.orilibrary.command.SubCommand

@SubCommand.Info(
    names = ["teleport", "go"],
    usage = "/island teleport [player]",
    permission = "skyblock.teleport"
)
class TeleportCommand(private val plugin: SkyblockPlugin) : SubCommand() {

    private val data = this.plugin.getManager<DataManager>()
    private val islandManager = this.plugin.getManager<IslandManager>()

    override fun executeArgument(sender: CommandSender, args: Array<String>) {

        // Check if the sender is a player.
        if (sender !is Player) {
            this.plugin.send(sender, "player-only")
            return
        }

        val member = this.data.getMember(sender.uniqueId)
        val island = data.getIsland(member.island)

        if (island == null) {
            this.plugin.send(sender, "no-island")
            return
        }

        this.islandManager.teleportToIsland(member, island)
    }
}